internal/adapter/lsp/server.go

```diff
@@ -389,6 +395,75 @@ func NewServer(opts ServerOpts) *Server {
 		return actions, nil
 	}
 
+	handler.TextDocumentReferences = func(context *glsp.Context, params *protocol.ReferenceParams) ([]protocol.Location, error) {
+		doc, ok := server.documents.Get(params.TextDocument.URI)
+		if !ok {
+			return nil, nil
+		}
+
+		link, err := doc.DocumentLinkAt(params.Position)
+		if link == nil || err != nil {
+			return nil, err
+		}
+
+		notebook, err := server.notebookOf(doc)
+		if err != nil {
+			return nil, err
+		}
+
+		target, err := server.noteForHref(link.Href, doc, notebook)
+		if link == nil || target == nil || err != nil {
+			return nil, err
+		}
+
+		opts := core.NoteFindOpts{}
+		p, err := notebook.RelPath(target.Path)
+		if err != nil {
+			return nil, err
+		}
+
+		opts.LinkTo = &core.LinkFilter{
+			Paths:       []string{p},
+			Negate:      false,
+			Recursive:   true,
+			MaxDistance: 1,
+		}
+
+		notes, err := notebook.FindNotes(opts)
+		if err != nil {
+			return nil, err
+		}
+
+		var locations []protocol.Location
+
+		for _, note := range notes {
+			pos := strings.Index(note.RawContent, target.Path[0:len(target.Path)-3])
```

Comment:
I expect this will fail if the notes are not in the same root folder, as the `target.Path` is always relative to the root of the notebook. The `-3` might be an issue too when not using file extensions.

However, worst case scenario is that we will preview the first line of the note, so I think it's fine for now. I'm planning to add an API to get the actual link infos which will provide the exact link position, we can refactor this then.

Reply:
It would be useful to have the actual position of the links in the files.

Reply:
Yep definitely. Right now in the database we have the position of the snippet containing the link, but I think I will add the link position itself as well.